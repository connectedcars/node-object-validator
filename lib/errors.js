function ValidationError(message, context = {}) {
  this.name = 'ValidationError'
  this.message = message
  if (context.key) {
    this.key = context.key
  }
  if (context.val) {
    this.val = context.val
  }
}
ValidationError.prototype = new Error()

module.exports = {
  ValidationError
}
