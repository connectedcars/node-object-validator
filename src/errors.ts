export interface ValidationErrorContext<T> {
  key: string
  value: T
}

export class ValidationError<T> extends Error {
  public key?: string
  public value?: T
  public constructor(message: string, context?: ValidationErrorContext<T>) {
    if (context) {
      message = `Field '${context.key}' ${message.charAt(0).toLowerCase() + message.slice(1)}`
    }
    super(message)
    this.key = context?.key
    this.value = context?.value
  }
}

export class RequiredError extends ValidationError<string> {}
export class NotStringError extends ValidationError<string> {}
export class WrongLengthError extends ValidationError<string> {}
export class NotDateError extends ValidationError<string> {}
export class NotRfc3339Error extends ValidationError<string> {}
export class NotExactStringError extends ValidationError<string> {}
export class NotFloatError extends ValidationError<string> {}
export class NotIntegerError extends ValidationError<string> {}
export class NotArrayError extends ValidationError<string> {}
export class NotObjectError extends ValidationError<string> {}
export class DoesNotMatchRegexError extends ValidationError<string> {}
export class OutOfRangeError extends ValidationError<string> {}
