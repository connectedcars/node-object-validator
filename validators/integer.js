const { ValidationError } = require('../lib/errors')

function Integer(min = Number.MIN_SAFE_INTEGER, max = Number.MAX_SAFE_INTEGER, required = true) {
  return {
    type: 'Integer',
    validate: function(key, val) {
      if (val == null) {
        if (required) {
          return new ValidationError(`Field \`${key}\` (${this.type}) is required`, { key, val })
        } else {
          return
        }
      }
      if (typeof val !== 'number' || !Number.isInteger(val)) {
        return new ValidationError(`Field \`${key}\` (${this.type}) must be an integer (received "${val}")`, {
          key,
          val
        })
      }
      if (val < min || val > max) {
        return new ValidationError(
          `Field \`${key}\` (${this.type}) must be between ${min} and ${max} (received "${val}")`,
          { key, val }
        )
      }
    }
  }
}

module.exports = Integer
