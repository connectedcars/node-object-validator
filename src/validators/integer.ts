import { NotIntegerError, OutOfRangeError, RequiredError, ValidationErrorContext } from '../errors'
import { Validator } from './common'

export function validateInteger(
  value: number,
  min = 0,
  max = Number.MAX_SAFE_INTEGER,
  context?: ValidationErrorContext
): Error | null {
  if (typeof value !== 'number' || !Number.isInteger(value)) {
    return new NotIntegerError(`Must be an integer (received "${value}")`, context)
  }
  if (value < min || value > max) {
    return new OutOfRangeError(`Must be between ${min} and ${max} (received "${value}")`, context)
  }
  return null
}

export class RequiredInteger implements Validator {
  private type: 'RequiredInteger' = 'RequiredInteger'
  private min: number
  private max: number

  public constructor(min = 0, max = Number.MAX_SAFE_INTEGER) {
    this.min = min
    this.max = max
  }

  public validate(value: number, context?: ValidationErrorContext): Error | null {
    if (value == null) {
      return new RequiredError(`Is required`, context)
    }
    return validateInteger(value, this.min, this.max, context)
  }
}

export class OptionalInteger implements Validator {
  private type: 'OptionalInteger' = 'OptionalInteger'
  private min: number
  private max: number

  public constructor(min = 0, max = Number.MAX_SAFE_INTEGER) {
    this.min = min
    this.max = max
  }

  public validate(value: number, context?: ValidationErrorContext): Error | null {
    if (value == null) {
      return null
    }
    return validateInteger(value, this.min, this.max, context)
  }
}

export function Integer(required?: false): OptionalInteger
export function Integer(required: true): RequiredInteger
export function Integer(required = false): OptionalInteger | RequiredInteger {
  return required ? new RequiredInteger() : new OptionalInteger()
}
