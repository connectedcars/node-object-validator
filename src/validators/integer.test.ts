import { NotIntegerFail, OutOfRangeFail, RequiredFail } from '../errors'
import { IntegerValidator, OptionalInteger, RequiredInteger, validateInteger } from './integer'

describe('Integer', () => {
  describe('validateInteger', () => {
    it('requires value to be an integer', function() {
      expect(validateInteger(0)).toStrictEqual([])
      expect(validateInteger(1)).toStrictEqual([])
      expect(validateInteger(123)).toStrictEqual([])
      expect(validateInteger('1')).toStrictEqual([new NotIntegerFail('Must be an integer (received "1")')])
      expect(validateInteger('')).toStrictEqual([new NotIntegerFail('Must be an integer (received "")')])
      expect(validateInteger({})).toStrictEqual([new NotIntegerFail('Must be an integer (received "[object Object]")')])
      expect(validateInteger([])).toStrictEqual([new NotIntegerFail('Must be an integer (received "")')])
      expect(validateInteger(true)).toStrictEqual([new NotIntegerFail('Must be an integer (received "true")')])
      expect(validateInteger(false)).toStrictEqual([new NotIntegerFail('Must be an integer (received "false")')])
    })

    it('requires min value', function() {
      expect(validateInteger(-1, 5, 500)).toStrictEqual([
        new OutOfRangeFail('Must be between 5 and 500 (received "-1")')
      ])
      expect(validateInteger(0, 5, 500)).toStrictEqual([new OutOfRangeFail('Must be between 5 and 500 (received "0")')])
      expect(validateInteger(1, 5, 500)).toStrictEqual([new OutOfRangeFail('Must be between 5 and 500 (received "1")')])
      expect(validateInteger(2, 5, 500)).toStrictEqual([new OutOfRangeFail('Must be between 5 and 500 (received "2")')])
      expect(validateInteger(3, 5, 500)).toStrictEqual([new OutOfRangeFail('Must be between 5 and 500 (received "3")')])
      expect(validateInteger(4, 5, 500)).toStrictEqual([new OutOfRangeFail('Must be between 5 and 500 (received "4")')])
      expect(validateInteger(5, 5, 500)).toStrictEqual([])
      expect(validateInteger(6, 5, 500)).toStrictEqual([])
      expect(validateInteger(123, 5, 500)).toStrictEqual([])
    })

    it('requires max value', function() {
      expect(validateInteger(-1, -500, 5)).toStrictEqual([])
      expect(validateInteger(0, -500, 5)).toStrictEqual([])
      expect(validateInteger(1, -500, 5)).toStrictEqual([])
      expect(validateInteger(2, -500, 5)).toStrictEqual([])
      expect(validateInteger(3, -500, 5)).toStrictEqual([])
      expect(validateInteger(4, -500, 5)).toStrictEqual([])
      expect(validateInteger(5, -500, 5)).toStrictEqual([])
      expect(validateInteger(6, -500, 5)).toStrictEqual([
        new OutOfRangeFail('Must be between -500 and 5 (received "6")')
      ])
      expect(validateInteger(7, -500, 5)).toStrictEqual([
        new OutOfRangeFail('Must be between -500 and 5 (received "7")')
      ])
    })
  })

  describe('IntegerValidator', () => {
    it('should generate validation code and give same result', () => {
      const validator = new IntegerValidator(1, 30, { optimize: true })
      const str = validator.validate.toString()
      expect(str).toMatch(/generatedFunction = true/)
      const errors = validator.validate(10)
      expect(errors).toEqual([])
    })
  })

  describe('RequiredInteger', () => {
    it('accepts empty value', function() {
      const validator = new RequiredInteger()
      expect(validator.validate(null)).toStrictEqual([new RequiredFail('Is required')])
      expect(validator.validate(undefined)).toStrictEqual([new RequiredFail('Is required')])
    })
  })

  describe('OptionalInteger', () => {
    it('accepts empty value', function() {
      const validator = new OptionalInteger()
      expect(validator.validate(null)).toStrictEqual([])
      expect(validator.validate(undefined)).toStrictEqual([])
    })
  })
})
