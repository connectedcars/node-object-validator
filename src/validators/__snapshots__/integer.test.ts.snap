// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integer RequiredInteger should return an function body 1`] = `
Array [
  Object {
    "NotIntegerFail": [Function],
    "OutOfRangeFail": [Function],
    "RequiredFail": [Function],
  },
  Array [],
  Array [
    "const value1 = value1",
    "if (value1 !== undefined) {",
    "  if (typeof value1 === 'number' && Number.isInteger(value1)) {",
    "    if (value1 < 0 || value1 > 10) {",
    "      errors.push(new OutOfRangeFail(\`Must be between 0 and 10\`, value1, context))",
    "    }",
    "  } else {",
    "    errors.push(new NotIntegerFail(\`Must be an integer\`, value1, context))",
    "  }",
    "} else {",
    "  errors.push(new RequiredFail(\`Is required\`, value1, context))",
    "}",
  ],
]
`;
